src/hammer-vlsi/par/openroad/__init__.py

```diff
@@ -84,148 +181,959 @@ def tool_config_prefix(self) -> str:
+                -rd out_file={self.output_gds_filename} \\
+                -rd tech_file={klayout_techfile} \\
+                -rm $::env(HAMMER_HOME)/src/hammer-vlsi/technology/sky130/extra/def2stream.py
+        """) 
+        # extra options:
+        #   -rd in_files="$::env(GDSOAS_FILES) $::env(WRAPPED_GDSOAS)" \: all the extra gds files (set to GDS_FILES += $(BLOCK_GDS), BLOCK_GDS += ./results/$(PLATFORM)/$(DESIGN_NICKNAME)_$(block)/$(FLOW_VARIANT)/6_final.gds)
+        #   -rd config_file=$fill_config \: ~OpenROAD-flow-scripts/flow/platforms/sky130hd/fill.json
+        #   -rd seal_file=$seal_gds \: I think we can skip??
+        return True
+
+    def write_sdf(self) -> bool:
+        corners = self.get_mmmc_corners()  
+        self.append(f"write_sdf -corner setup {self.output_sdf_path}")
+        return True
+
+    def write_spefs(self) -> bool:
+        
         return True
 
     def write_regs(self) -> bool:
         # TODO: currently no analagous OpenROAD default script
         return True
 
+    def write_reports(self) -> bool:
+        # TODO: write all these to files!!!! (or process log to generate report)
+        self.block_append("""
+        ################################################################
+        # Final Report
+        report_checks -path_delay min_max -format full_clock_expanded -fields {input_pin slew capacitance} -digits 3
+        report_worst_slack -min -digits 3
+        report_worst_slack -max -digits 3
+        report_tns -digits 3
+        report_check_types -max_slew -max_capacitance -max_fanout -violators -digits 3
+        report_clock_skew -digits 3
+        report_power -corner "hold"
+
+        report_floating_nets -verbose
+        report_design_area
+
+        utl::metric "worst_slack_min" [sta::worst_slack -min]
+        utl::metric "worst_slack_max" [sta::worst_slack -max]
+        utl::metric "tns_max" [sta::total_negative_slack -max]
+        utl::metric "clock_skew" [sta::worst_clock_skew -setup]
+        utl::metric "max_slew_violations" [sta::max_slew_violation_count]
+        utl::metric "max_fanout_violations" [sta::max_fanout_violation_count]
+        utl::metric "max_capacitance_violations" [sta::max_capacitance_violation_count]
+        # report clock period as a metric for updating limits
+        utl::metric "clock_period" [get_property [lindex [all_clocks] 0] period]
+        """)
+        return True
+
     def write_design(self) -> bool:
-        # TODO: currently using OpenROAD's default finish script
-        self.cmds += [
-          "make DESIGN_CONFIG={conf} -f {make} finish".format(
-            conf=self.design_config_path(), 
-            make=self.openroad_flow_makefile_path()
-        )]
+        self.block_append(f"""
+        ################################################################
+        # Write Design
+        """)
+        self.append(f"write_def {self.output_def_filename}")
```

Comment:
If klayout needs DEFs for writing GDS, isn't this one the only one that's necessary? Any advantage to writing checkpoint DEFs?

Reply:
took these out, they were just in the original flow